### PR TITLE
Feature/19538 add upload to live agent through webchat

### PR DESCRIPTION
**Pull request by cbuchwald**
This PR adds support for file uploading to the Live Agent.

How to test:

1. Create a conversation with handover provider Live Agent and get sure you can exchange messages between user and agent
2. Start your chrome browser in unsafe mode to avoid CORS issues with 'google-chrome --disable-web-security --user-data-dir="~"'
3. Configure your webchat widget to use a fixed userId and sessionId and chat with the agent. Check requests in the network tab to retrieve your URLToken
4. Build and import the file-upload plugin from this package after the webchat.js or webchat.legacy.js file in the file you use to call the widget.
5. Either fork or export and import the Chatwoot APIs from postman https://www.postman.com/chatwoot/workspace/chatwoot-apis/overview
6. Create/update the following environment variables:
- host = https://dev-live-agent.cognigy.ai
- api_version = api/v1
- api_access_token = Access Token from your Live Agent Profile Settings
- account_id = Your Live Agent Account Id, can be taken from the URL
7. Call "Collections-->Chatwoot Application API v1.0-->Conversations-->List all conversations".
8. Search the JSON output from the previous step. Add a new environment variable called contact_id and assign it the value from  data.payload[0].meta.sender.id
9. Call "Collections-->Chatwoot Application API v1.0-->Contacts-->Get Contact" and note the value from payload.contact_inboxes[0].source_id as client_identifier.
10. Use the notify API from service-api and send such a request to your conversation:

{
  "userId": "1", //<--same as in your webchat widget config
  "text": "",
  "data": {
"_plugin": {
"type": "file-upload",
"service": "live-agent",
"inboxIdentifier":"RC1r4dqhQxiVHQrpQjHZKoxm",//<-- taken from the configuration of your Live Agent inbox(not the number in the URL)
"accountId":"7", //<--your user account in Live Agent, can be taken from the URL
"conversationId":"28", //<--your conversationId in Live Agent, can be taken from the URL
"contactIdentifier":"xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx", //<--retrieved in step 9
"host":"https://dev-live-agent.cognigy.ai"}}, //<--the hostname of your Live Agent installation
  "URLToken": "xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx", //<--taken from network tab while in conversation
  "sessionId": "1" //<--same as in your webchat widget config
}

11. You should now receive a file upload button that you can press for uploading. Select a picture and it should get uploaded.
12. The picture should be available in the Live Agent afterwards
13. Upload files with these extensions: .png, jpg, jpeg, pdf, all should be rendered as images, except for the pdf, it should have a download button.

Our Live Agent installation is facing some issues when downloading files using the url we received, therefore I discussed with Nacho that this is out of scope and will be solved by his team

### CODE REVIEW

**Comment by pedily** [comment]
I tried uploading a png and a JPG file but they didn't show up in the Live Agent application. Instead, I got text messages saying "file-upload". In addition, those messages will also get forwarded to the webchat user. As soon as i "uploaded a file" (HTTP request was successful), I will see 2 new messages:
- a text message from "the agent" saying "file-upload" (unexpected, shouldn't be visible for both me and the agent)
- an image message from "the agent" that does not get loaded (unexpected as well, this should be visible for the agent but not for me)
Screenshot from 2021-11-26 11-51-52
Screenshot from 2021-11-26 11-55-18

I think the "images don't load" thing could be related to the issue you mentioned, but we still need to hide the "file upload" message for both parties and the "the file you sent" message for the webchat user

**Comment by pedily** [comment]
We need to find an alternate authentication mechanism other than publishing the live agent user's apiKey. This is a security risk as any user could use this token to do anything live agent related on the live agent user's behalf.

@Nachox07 is there any other way to authenticate this, e.g. using a  single-use "throwaway token" that's tied to a certain action?
Otherwise we will need to solve this differently, e.g. by building a proxy into our Endpoint

File: plugins/file-upload/src/helpers/upload.js [hunk 1]
```diff
@@ -15,14 +18,30 @@ export const upload = async (config, file) => {
...
+                .then(() => downloadUrl)
         }
+        case 'live-agent': {
+            const { userApiKey, host, accountId, conversationId } = config;
+            const uploadUrl = `${host}/api/v1/accounts/${accountId}/conversations/${conversationId}/messages`;
+            const form = new FormData();
+            form.append('attachments[]', file);
+            form.append('content', 'file-upload');
+
+            return Axios.post(uploadUrl, form, {
```
Comment (Nachox07) [inline plugins/file-upload/src/helpers/upload.js:38]: Is it necessary to use Axios instead of fetch?

Comment (cbuchwald) [reply plugins/file-upload/src/helpers/upload.js:38]: I tried doing it with fetch for hours and unfortunately it didn't work, Axios is the only way it works even if all parameters are the same

**Review by Nachox07 — APPROVED** [review]

**Comment by cbuchwald** [comment]
> I tried uploading a png and a JPG file but they didn't show up in the Live Agent application. Instead, I got text messages saying "file-upload". In addition, those messages will also get forwarded to the webchat user. As soon as i "uploaded a file" (HTTP request was successful), I will see 2 new messages:
> 
> * a text message from "the agent" saying "file-upload" (unexpected, shouldn't be visible for both me and the agent)
> * an image message from "the agent" that does not get loaded (unexpected as well, this should be visible for the agent but not for me)
>   Screenshot from 2021-11-26 11-51-52
>   Screenshot from 2021-11-26 11-55-18
> 
> I think the "images don't load" thing could be related to the issue you mentioned, but we still need to hide the "file upload" message for both parties and the "the file you sent" message for the webchat user

Unfortunately Live Agent requires a parameter called content with a string so I entered "file-upload". I can't omit it, but I will change it to "file successfully uploaded". I guess it will be visible on user side when the bug with the URLs of uploaded items is fixed in Live Agent, therefore I would consider it out of scope as showing it to the user wasn't part of the requirements

**Comment by cbuchwald** [comment]
> We need to find an alternate authentication mechanism other than publishing the live agent user's apiKey. This is a security risk as any user could use this token to do anything live agent related on the live agent user's behalf.
> 
> @Nachox07 is there any other way to authenticate this, e.g. using a single-use "throwaway token" that's tied to a certain action? Otherwise we will need to solve this differently, e.g. by building a proxy into our Endpoint

fixed by using the contact identifier instead

**Review by Nachox07 — APPROVED** [review]

**Review by Cesarcs7 — APPROVED** [review]
Works for me, I have tested it with my Live Agent implementation.

**Review by pedily — APPROVED** [review]
Works as expected. Images don't currently render in Live Agent, but that's a separate issue @Nachox07

**Pull request merged**